lib/libvfio-user.c

```diff
@@ -2273,7 +2273,15 @@ vfu_sgl_write(vfu_ctx_t *vfu_ctx, dma_sg_t *sgl, size_t cnt, void *data)
         return ERROR_INT(ENOTSUP);
     }
 
-    return vfu_dma_transfer(vfu_ctx, VFIO_USER_DMA_WRITE, sgl, data);
+    int ret = vfu_dma_transfer(vfu_ctx, VFIO_USER_DMA_WRITE, sgl, data);
+
+    if (ret < 0) {
+        return ret;
+    }
+
+    vfu_sgl_mark_dirty(vfu_ctx, sgl, cnt);
```

Comment:
nit: I guess you can just:
```
vfu_sgl_mark_dirty(vfu_ctx, sgl, cnt);
return vfu_dma_transfer(vfu_ctx, VFIO_USER_DMA_WRITE, sgl, data);
```
I don't think it's a problem marking something dirty when it's actually clean in case of error?

Reply:
The marking has to happen after the write finishes.

Reply:
Oh right otherwise the _client_ can race.

Reply:
I was referring to this race:

1. page is marked as dirty
2. dirty page get returns bitmap, clears bit
3. page is copied into client
4. client receives DMA write request, writes to page, but it's no longer known to be dirty


But now I'm starting to think maybe the right thing is for the *client* to handle the dirtiness. It is, after all, the one doing the writes. (Sorry @w-henderson ).